app/src/org/commcare/engine/cases/CaseUtils.java

```diff
@@ -79,38 +80,12 @@ public static String computeCaseDbHash(SqlStorage<?> storage) {
     public static void purgeCases() throws InvalidCaseGraphException {
         long start = System.currentTimeMillis();
         //We need to determine if we're using ownership for purging. For right now, only in sync mode
-        Vector<String> owners = new Vector<>();
-        Vector<String> users = new Vector<>();
-        for (IStorageIterator<User> userIterator = CommCareApplication.instance()
-                .getUserStorage(User.STORAGE_KEY, User.class).iterate(); userIterator.hasMore(); ) {
-            String id = userIterator.nextRecord().getUniqueId();
-            owners.addElement(id);
-            users.addElement(id);
-        }
-
-        //Now add all of the relevant groups
-        //TODO: Wow. This is.... kind of megasketch
-        for (String userId : users) {
-            DataInstance instance = CommCareUtil.loadFixture("user-groups", userId);
-            if (instance == null) {
-                continue;
-            }
-            EvaluationContext ec = new EvaluationContext(instance);
-            for (TreeReference ref : ec.expandReference(XPathReference.getPathExpr("/groups/group/@id").getReference())) {
-                AbstractTreeElement<AbstractTreeElement> idelement = ec.resolveReference(ref);
-                if (idelement.getValue() != null) {
-                    owners.addElement(idelement.getValue().uncast().getString());
-                }
-            }
-        }
-
-        SQLiteDatabase db;
-        db = CommCareApplication.instance().getUserDbHandle();
-
-        db.beginTransaction();
+        Vector<String> owners = getAllOwners();
+        SQLiteDatabase db = CommCareApplication.instance().getUserDbHandle();
         int removedCaseCount;
         int removedLedgers;
         try {
+            db.beginTransaction();
```

Comment:
I believe our pattern is for these to stay [outside of the transaction](https://github.com/dimagi/commcare-android/blob/d3f29f1cc1bce0fdaf14d8c5d4ce00f64c464c6d/app/src/org/commcare/xml/AndroidBulkCaseXmlParser.java#L61) when being used as `begin/success/end`, otherwise if this method itself throws an error and the transaction wasn't started, the `endTransaction` call could unintentionally end a separate transaction instead of this one.

Reply:
https://github.com/dimagi/commcare-android/pull/2955/commits/78f2e5ddd95df21034763976fe824899f6395232